commons-geometry-examples/examples-jmh/src/main/java/org/apache/commons/geometry/examples/jmh/euclidean/VectorPerformance.java

```diff
@@ -0,0 +1,382 @@
+    private static final String EDGE = "edge";
+
+    /** Base class for vector inputs.
+     * @param <V> Vector implementation type
+     */
+    @State(Scope.Thread)
+    public abstract static class VectorInputBase<V extends Vector<V>> {
+
+        /** The dimension of the vector. */
+        private final int dimension;
+
+        /** Factory function used to create vectors from arrays of doubles. */
+        private final Function<double[], V> vectorFactory;
+
+        /** The number of vectors in the input list. */
+        @Param({"1000"})
+        private int size;
+
+        /** The vector for the instance. */
+        private List<V> vectors;
+
+        /** Create a new instance with the vector dimension.
+         * @param dimension vector dimension
+         * @param vectorFactory function for creating vectors from double arrays
+         */
+        VectorInputBase(final int dimension, final Function<double[], V> vectorFactory) {
+            this.dimension = dimension;
+            this.vectorFactory = vectorFactory;
+        }
+
+        /** Set up the instance for the benchmark.
+         */
+        @Setup(Level.Iteration)
+        public void setup() {
+            vectors = new ArrayList<>(size);
+
+            final double[] values = new double[dimension];
+            final DoubleSupplier doubleSupplier = createDoubleSupplier(getType(),
+                    RandomSource.create(RandomSource.XO_RO_SHI_RO_128_PP));
+
+            for (int i = 0; i < size; ++i) {
+                for (int j = 0; j < dimension; ++j) {
+                    values[j] = doubleSupplier.getAsDouble();
+                }
+
+                vectors.add(vectorFactory.apply(values));
+            }
+        }
+
+        /** Get the input vectors for the instance.
+         * @return the input vectors for the instance
+         */
+        public List<V> getVectors() {
+            return vectors;
+        }
+
+        /** Get the type of double values to use in the creation of input vectors.
+         * @return the type of double values to use in the creation of input vectors
+         */
+        public abstract String getType();
+
+        /** Create a supplier that produces doubles of the given type.
+         * @param type type of doubles to produce
+         * @param rng random provider
+         * @return a supplier that produces doubles of the given type
+         */
+        private DoubleSupplier createDoubleSupplier(final String type, final UniformRandomProvider rng) {
+            switch (type) {
+            case RANDOM:
+                return () -> createRandomDouble(rng);
+            case NORMALIZABLE:
+                final ZigguratNormalizedGaussianSampler sampler = ZigguratNormalizedGaussianSampler.of(rng);
+                return () -> {
+                    double n = sampler.sample();
+                    return Math.abs(n) == 0 ? 0.1 : n; // do not return exactly zero
```

Comment:
No need for Math.abs here, -0.0 or 0.0 will equal 0.

I assume this is to guard against vectors that cannot be normalised as they sum to 0. In this case I would also guard the RANDOM ones since they could also be zero, however unlikely.

Reply:
Got it. Yes, this is simply to ensure that the Vector.normalize() benchmarks do not throw due to input vectors with a zero norm. I did not include this check in the RANDOM input type because other methods (ie, Vector.norm()) do accept zero-length vectors.